### PR TITLE
Site Settings final pass - design, analytics, feature toggling

### PR DESCRIPTION
**Pull request by tonyr59h**
This PR will cover #3434, #3441, and #3477.

~~Not ready for developer review, opening this for @mattmiklic to verify that the design is correct and to visualize the diff for myself.~~

Edit:
All the issues should be covered now, ready for review!

### CODE REVIEW

**Comment by aforcier** [comment]
Following our conversation, I gave the number picker a quick test on an `API 18` Samsung S3 to check for OEM landmines on their part (since reflection was used for styling and the internals may have been changed).

Everything seems to work, no crashes. :+1:

There's only a border issue that's likely API-level related and not specific to the Samsung:

samsung-s3-reflection-picker

**Comment by mattmiklic** [comment]
It's looking good now; most of my comments are fairly minor at this point. Roughly in order of importance:

On all dialogs that include just a text input (site title, tagline, adding items to the moderation/blacklist, etc.); when the dialog appears the input is focused but the keyboard isn't visible; you have to tap the input to bring up the keyboard. Could we show the keyboard automatically?

On the Hold for Moderation and Blacklist screens, if I select two or more items and tap the Delete icon, only one item is deleted.

On screens where we include a picker, ideally the borders will appear directly above and below the picker, separating it from any other text or header in the dialog. Here are two examples (these screens also have a few font issues I have noted in red):

screen shot 2015-12-08 at 12 15 01 pm
screen shot 2015-12-08 at 12 21 40 pm

The Discussion settings screen has a few items that are in Roboto instead of Open Sans, and the border color is off.

screen shot 2015-12-08 at 12 26 52 pm

**Comment by mattmiklic** [comment]
I just noticed one bigger issue; "Close after" is missing the toggle switch to turn that setting on or off.

screen shot 2015-12-08 at 12 53 32 pm

**Comment by tonyr59h** [comment]
Thanks for your feedback @daniloercoli, all your points are addressed in 9383f4f7c680b335f52f4fe3266b1b67892325fa

The design has been updated based on your response @mattmiklic. The one issue I haven't fixed is the Discussion dialog dividers, it's still on my list though :+1:

@kwonye all the functionality is in now. It's ready for a developer review :)

**Comment by daniloercoli** [comment]
Not related to this PR, but I'm getting a crash in site settings for a .com test site here:
https://github.com/wordpress-mobile/WordPress-Android/blob/9383f4f7c680b335f52f4fe3266b1b67892325fa/WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java#L310

`params.put(MODERATION_KEYS_KEY, builder.substring(0, builder.length() - 1));`

`builder.length() ` is equal to 1 in my case.

**Comment by daniloercoli** [comment]
Also getting a crash If I change a site setting and then immediately go back to the main site screen. Seems that the `error` handler doesn't check if the fragment is still attached to the activity.
```
12-10 10:03:03.125 5048-5048/org.wordpress.android E/AndroidRuntime: FATAL EXCEPTION: main
                                                                     Process: org.wordpress.android, PID: 5048
                                                                     java.lang.IllegalStateException: Fragment SiteSettingsFragment{c92018c} not attached to Activity
                                                                         at android.app.Fragment.getResources(Fragment.java:788)
                                                                         at org.wordpress.android.ui.prefs.SiteSettingsFragment.setPreferencesFromSiteSettings(SiteSettingsFragment.java:625)
                                                                         at org.wordpress.android.ui.prefs.SiteSettingsFragment.onSettingsUpdated(SiteSettingsFragment.java:383)
                                                                         at org.wordpress.android.ui.prefs.SiteSettingsInterface$1.run(SiteSettingsInterface.java:501)
                                                                         at android.app.Activity.runOnUiThread(Activity.java:5247)
                                                                         at org.wordpress.android.ui.prefs.SiteSettingsInterface.notifyUpdatedOnUiThread(SiteSettingsInterface.java:498)
                                                                         at org.wordpress.android.ui.prefs.DotComSiteSettings$5.onResponse(DotComSiteSettings.java:341)
                                                                         at org.wordpress.android.ui.prefs.DotComSiteSettings$5.onResponse(DotComSiteSettings.java:329)
                                                                         at com.wordpress.rest.RestRequest.deliverResponse(RestRequest.java:74)
                                                                         at com.wordpress.rest.RestRequest.deliverResponse(RestRequest.java:18)
                                                                         at com.android.volley.ExecutorDelivery$ResponseDeliveryRunnable.run(ExecutorDelivery.java:99)
                                                                         at android.os.Handler.handleCallback(Handler.java:739)
                                                                         at android.os.Handler.dispatchMessage(Handler.java:95)
                                                                         at android.os.Looper.loop(Looper.java:135)
                                                                         at android.app.ActivityThread.main(ActivityThread.java:5221)
                                                                         at java.lang.reflect.Method.invoke(Native Method)
                                                                         at java.lang.reflect.Method.invoke(Method.java:372)
                                                                         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:899)
                                                                         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:694)
```

**Comment by tonyr59h** [comment]
> Not related to this PR, but I'm getting a crash in site settings for a .com test site here

I'd say it's related ;) Fixed in 4b9a7e09fef44ca093a47726e1b37a6f77bac26d

**Comment by tonyr59h** [comment]
> Also getting a crash If I change a site setting and then immediately go back to the main site screen. 

I think 678ce0864c382b81e82fae6f34cb14c77be8b014 should address this one. The Activity is checked to make sure it's not finishing before calling the listener and the fragment makes sure it's added before updating.

**Comment by mattmiklic** [comment]
Hey Tony, just took a look and have a few things to mention:

- In Hold for Moderation, after selecting multiple items, when I tapped the Delete/trash icon I got a crash (that I can reproduce)
- In Hold for Moderation and Blacklist, the keyboard doesn't appear automatically when the dialog opens.
- In Site Settings and Tagline, the keyboard doesn't disappear automatically when the dialog is dismissed.
- In Close commenting, "Never" shouldn't be an option in the picker (commenting never closes if the toggle switch on that screen is turned off).
- On the threading screen, we need to add options for 6, 7, 8, 9, or 10 levels of threading, so it offers the same options as the web (this is a change from my design). Right now we only offer up to five, but I noticed that if I set it to 6 or higher on the web, it causes a crash in the app.
- Tapping "Links in Comments" crashes the app.
- Buttons on the Close commenting, Paging screen are in Roboto instead of Open Sans.

All of this is on a Nexus 5 (let me know if you need any more info on my setup for the crashes).

File: libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTracker.java [hunk 1]
```diff
@@ -120,9 +119,18 @@
         THEMES_CUSTOMIZE_ACCESSED,
         THEMES_SUPPORT_ACCESSED,
         THEMES_DETAILS_ACCESSED,
+        SETTINGS_LANGUAGE_SELECTION_FORCED,
```
Comment (daniloercoli) [inline libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTracker.java]: Should this be changed to `SITE_SETTINGS_LANGUAGE_SELECTION_FORCED` ?

File: libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java [hunk 2]
```diff
@@ -375,6 +372,33 @@ public void track(AnalyticsTracker.Stat stat, Map<String, ?> properties) {
             case THEMES_DETAILS_ACCESSED:
                 eventName = "themes_details_accessed";
                 break;
+            case SETTINGS_LANGUAGE_SELECTION_FORCED:
+                eventName = "settings_language_selection_forced";
```
Comment (daniloercoli) [inline libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java]: If we change the KEY to `SITE_SETTINGS_LANGUAGE_SELECTION_FORCED` we need to change the value here.

File: libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTracker.java (hunk 1 above)
Comment (tonyr59h) [inline libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTracker.java]: That option is not part of the Site Settings, it's under Account Settings. I could change it to `ACCOUNT_SETTINGS_LANGUAGE_SELECTION_FORCED`.

**Comment by kwonye** [comment]
Regarding bcee50e, are we effectively disabling settings for self-hosted (4.5 isn't out yet)? If so, I'd like to see the settings field disabled instead of what it does now, which is just kick you out of the screen immediately and say there was an error.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java [hunk 3]
```diff
@@ -87,6 +90,22 @@ public void onResponse(JSONObject response) {
                         AppLog.d(AppLog.T.API, "Site Settings saved remotely");
                         notifySavedOnUiThread(null);
                         mRemoteSettings.copyFrom(mSettings);
+
+                        if (response != null) {
+                            JSONObject updated = response.optJSONObject("updated");
```
Comment (kwonye) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java]: hard-coded string

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java [hunk 4]
```diff
@@ -87,6 +90,22 @@ public void onResponse(JSONObject response) {
...
+                            if (updated == null) return;
+                            HashMap<String, Object> properties = new HashMap<>();
+                            Iterator<String> keys = updated.keys();
+                            while (keys.hasNext()) {
+                                String currentKey = keys.next();
+                                Object currentValue = updated.opt(currentKey);
+                                if (currentValue != null) {
+                                    properties.put("item_saved_" + currentKey, currentValue);
```
Comment (kwonye) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java]: hard-coded string

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SelfHostedSiteSettings.java [hunk 5]
```diff
@@ -86,6 +94,22 @@ public void saveSettings() {
...
+
+                if (result != null) {
+                    HashMap<String, Object> properties = new HashMap<>();
+                    if (result instanceof Map) {
+                        Map<String, Object> resultMap = (Map) result;
+                        Set<String> keys = resultMap.keySet();
+                        for (String key : keys) {
+                            Object currentValue = resultMap.get(key);
+                            if (currentValue != null) {
+                                properties.put("item_saved_" + key, currentValue);
```
Comment (kwonye) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SelfHostedSiteSettings.java]: hard-coded string

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SelfHostedSiteSettings.java [hunk 6]
```diff
@@ -151,6 +175,11 @@ public void onFailure(long id, Exception error) {
         public void onSuccess(long id, final Object result) {
             if (result instanceof Map) {
                 AppLog.d(AppLog.T.API, "Received Options XML-RPC response.");
+
+                if (!versionSupported((Map) result)) {
+                    notifyUpdatedOnUiThread(new XMLRPCException("Unsupported WordPress version"));
```
Comment (kwonye) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SelfHostedSiteSettings.java]: Please add to `strings.xml`

File: libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTracker.java (hunk 1 above)
Comment (daniloercoli) [inline libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTracker.java]: Yes please, rename it. :)

**Comment by kwonye** [comment]
- There seems to be an issue with `WPSwitch`:
  - Can't turn on related posts
  - Can't disable "Close after" in comments
  - Can't disable "paging" in comments doesn't do anything.

- Default format is still not showing up even for WordPress.com sites. Is this a regression?

screenshot_20151212-125201
screenshot_20151212-125205

**Comment by tonyr59h** [comment]
> Can't turn on related posts / Default format is still not showing up even for WordPress.com sites

Both of these are working on my end. I'll sync up tomorrow morning so we can go through it. In the meantime can you post your test device specs?

**Comment by kwonye** [comment]
> Both of these are working on my end. I'll sync up tomorrow morning so we can go through it. In the meantime can you post your test device specs?

Interesting! I'd imagine others didn't have issues either because they didn't report it. I'll investigate.

Device was Nexus 6P, 6.0

**Comment by tonyr59h** [comment]
Good catch with the dialog switches @kwonye. Are you still seeing the default format not set originally? I've tried on two sites and can't reproduce. For the first site I did a fresh install to make sure there was no cached data.

**Comment by kwonye** [comment]
> Are you still seeing the default format not set originally? 
Yes, @mattmiklic and @daniloercoli do you guys mind seeing if it works on your end please?

When opening settings, the Default Format is not set.

**Comment by kwonye** [comment]
Also, for sites where I do not have access, do not show settings. Right now it is shown and it just kicks me out saying "couldn't retrieve site info". (Only sites where I am an administrator can settings be changed)

It is like this on the web: (Here I am an editor)
screen shot 2015-12-14 at 1 51 24 pm

**Comment by tonyr59h** [comment]
> Also, for sites where I do not have access, do not show settings.

Good idea! Added in 8de264efe19c4620f983d8cf2de6a6ed57fc64f6

**Comment by tonyr59h** [comment]
@kwonye I just went through a full-pass on WP.com, starting with a fresh install. I'm still unable to reproduce the default post format bug you're seeing. Does it save any changes you make to it? If so then I don't think we should consider it a blocking issue, barring anyone else seeing it.

I tested with HTC One M8, Android 5.1.

**Comment by daniloercoli** [comment]
Default post formats crash
```
12-15 09:33:42.978 3224-3224/org.wordpress.android E/AndroidRuntime: FATAL EXCEPTION: main
                                                                     Process: org.wordpress.android, PID: 3224
                                                                     java.lang.NullPointerException: Attempt to invoke interface method 'java.lang.Object java.util.Map.get(java.lang.Object)' on a null object reference
                                                                         at org.wordpress.android.ui.prefs.SiteSettingsInterface.getDefaultPostFormatDisplay(SiteSettingsInterface.java:218)
                                                                         at org.wordpress.android.ui.prefs.SiteSettingsFragment.onPreferenceChange(SiteSettingsFragment.java:400)
                                                                         at android.preference.Preference.callChangeListener(Preference.java:928)
                                                                         at org.wordpress.android.ui.prefs.DetailListPreference.access$500(DetailListPreference.java:30)
                                                                         at org.wordpress.android.ui.prefs.DetailListPreference$DetailListAdapter.changeSelection(DetailListPreference.java:279)
                                                                         at org.wordpress.android.ui.prefs.DetailListPreference$DetailListAdapter.access$400(DetailListPreference.java:217)
                                                                         at org.wordpress.android.ui.prefs.DetailListPreference$DetailListAdapter$2.onClick(DetailListPreference.java:266)
                                                                         at android.view.View.performClick(View.java:4756)
                                                                         at android.view.View$PerformClick.run(View.java:19749)
                                                                         at android.os.Handler.handleCallback(Handler.java:739)
                                                                         at android.os.Handler.dispatchMessage(Handler.java:95)
                                                                         at android.os.Looper.loop(Looper.java:135)
                                                                         at android.app.ActivityThread.main(ActivityThread.java:5221)
                                                                         at java.lang.reflect.Method.invoke(Native Method)
                                                                         at java.lang.reflect.Method.invoke(Method.java:372)
                                                                         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:899)
                                                                         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:694)
```

**Comment by mattmiklic** [comment]
Took a look at the latest and it's looking pretty good. Noticed a few things:

- If I delete all of the entries in the Hold for moderation list, nothing is actually deleted on the server. If I delete them one at a time, it works correctly.
- Border colors in Discussion settings need to be updated to match the main Settings screen.
- Hold for Moderation/Blacklist is only saved when you go back to the Discussion settings screen. It would be nice to save immediately when something is added or removed from the list, to match the behavior elsewhere.

I tried toggling every setting on my dotcom test blog and it worked great. The speed of saving is really nice now that it happens immediately for each option.

**Comment by tonyr59h** [comment]
> The perception of speed is really enhanced now that it saves immediately for each option.

Awesome :) Now to mimic that sense of speed on self-hosted...

Good testing @daniloercoli, I never saw that crash. It's probably best to disable all preferences until we get all the remote data.

**Comment by tonyr59h** [comment]
Recent commits address all of Matt's comments. @kwonye ready for another go.

**Comment by daniloercoli** [comment]
Getting an issue with `Default Category`.  The modal could be empty under some circumstances (slow network connection):
screen shot 2015-12-16 at 09 40 40

**Comment by daniloercoli** [comment]
Another round of testing on `getPostFormats`.
The app does a 1st call to `wp.getPostFormats` as soon as the user switches site in the site picker, or when synching blog content in background/AsynchTask. In this 1st case the XML-RPC call made to the server is the following:
```
<methodCall>
	<methodName>wp.getPostFormats</methodName>
	<params>
		<param>
			<value>
				<i4>15592836</i4>
			</value>
		</param>
		<param>
			<value>
				<string>daniloercoli</string>
			</value>
		</param>
		<param>
			<value>
				<string />
			</value>
		</param>
		<param>
			<value>
				<string>show-supported</string>
			</value>
		</param>
	</params>
</methodCall>
```
Note that the parameter `show-supported` is missing from its value.
The response is the following:
```
<methodResponse> 
	<params> 
		<param> 
			<value> 
				<struct> 
					<member>
						<name>standard</name>
						<value>
							<string>Standard</string>
						</value>
					</member> 
					<member>
						<name>aside</name>
						<value>
							<string>Aside</string>
						</value>
					</member> 
					<member>
						<name>chat</name>
						<value>
							<string>Chiacchierare</string>
						</value>
					</member> 
					<member>
						<name>gallery</name>
						<value>
							<string>Gallery</string>
						</value>
					</member> 
					<member>
						<name>link</name>
						<value>
							<string>Link</string>
						</value>
					</member> 
					<member>
						<name>image</name>
						<value>
							<string>Immagine</string>
						</value>
					</member> 
					<member>
						<name>quote</name>
						<value>
							<string>Quote</string>
						</value>
					</member> 
					<member>
						<name>status</name>
						<value>
							<string>Status</string>
						</value>
					</member> 
					<member>
						<name>video</name>
						<value>
							<string>Video</string>
						</value>
					</member> 
					<member>
						<name>audio</name>
						<value>
							<string>Audio</string>
						</value>
					</member> 
				</struct> 
			</value> 
		</param> 
	</params> 
</methodResponse>
```
The server replied ignoring the `show-supported` parameter.

The 2nd call to `wp.getPostFormats` is made when the user taps on the Settings menu item.
In this case the call is correctly made to the server, including the value for the parameter `show-supported`.
```
<methodCall>
	<methodName>wp.getPostFormats</methodName>
	<params>
		<param>
			<value>
				<i4>15592836</i4>
			</value>
		</param>
		<param>
			<value>
				<string>daniloercoli</string>
			</value>
		</param>
		<param>
			<value>
				<string />
			</value>
		</param>
		<param>
			<value>
				<struct>
					<member>
						<name>show-supported</name>
						<value>
							<string>true</string>
						</value>
					</member>
				</struct>
			</value>
		</param>
	</params>
</methodCall>
```

And the response is the following:
```
<methodResponse> 
	<params> 
		<param> 
			<value> 
				<struct> 
					<member>
						<name>all</name>
						<value>
							<struct> 
								<member>
									<name>standard</name>
									<value>
										<string>Standard</string>
									</value>
								</member> 
								<member>
									<name>aside</name>
									<value>
										<string>Aside</string>
									</value>
								</member> 
								<member>
									<name>chat</name>
									<value>
										<string>Chiacchierare</string>
									</value>
								</member> 
								<member>
									<name>gallery</name>
									<value>
										<string>Gallery</string>
									</value>
								</member> 
								<member>
									<name>link</name>
									<value>
										<string>Link</string>
									</value>
								</member> 
								<member>
									<name>image</name>
									<value>
										<string>Immagine</string>
									</value>
								</member> 
								<member>
									<name>quote</name>
									<value>
										<string>Quote</string>
									</value>
								</member> 
								<member>
									<name>status</name>
									<value>
										<string>Status</string>
									</value>
								</member> 
								<member>
									<name>video</name>
									<value>
										<string>Video</string>
									</value>
								</member> 
								<member>
									<name>audio</name>
									<value>
										<string>Audio</string>
									</value>
								</member> 
							</struct>
						</value>
					</member> 
					<member>
						<name>supported</name>
						<value>
							<array>
								<data> 
									<value>
										<string>aside</string>
									</value> 
									<value>
										<string>image</string>
									</value> 
									<value>
										<string>video</string>
									</value> 
									<value>
										<string>link</string>
									</value> 
									<value>
										<string>quote</string>
									</value> 
									<value>
										<string>gallery</string>
									</value> 
									<value>
										<string>status</string>
									</value> 
									<value>
										<string>audio</string>
									</value> 
								</data>
							</array>
						</value>
					</member> 
				</struct> 
			</value> 
		</param> 
	</params> 
</methodResponse>
```

I bet we can unify the 2 calls above, and reuse the response from the first call is already there, and speed up the first setup of the site settings UI. (We already have the postFormats in the db and the call to upgrade it could be postpone in favor of other settings calls).

**Comment by kwonye** [comment]
- Self-hosted settings are still popping out with "Couldn't retrieve site info"
- Was @daniloercoli's comment addressed?

**Comment by kwonye** [comment]
self-hosted issue resolved with #3562

**Comment by tonyr59h** [comment]
> Was @daniloercoli's comment addressed?

Addressed in 46eb536402d80744ed7ba3221ab68b98850c4759

**Comment by kwonye** [comment]
Manual testing all looks good! :shipit:

**Pull request merged**